doc/source/reference/c-api.dtype.rst

```diff
@@ -374,3 +374,30 @@ format specifier in printf and related commands.
     :c:data:`NPY_LONGLONG_FMT`, :c:data:`NPY_ULONGLONG_FMT`,
     :c:data:`NPY_INTP_FMT`, :c:data:`NPY_UINTP_FMT`,
     :c:data:`NPY_LONGDOUBLE_FMT`
+
+Generic Typing in Source Files
+------------------------------
+
+Many of the C source files (extension: :file:`.c.src`) in NumPy are written
+in a custom templating language that is used to generate C code.
+The :c:data:`@` symbol is used to wrap macro-style variables to empower
+a substitution mechansim for the large number of data types described above.
+For example, a loop in the NumPy C source code may have a :c:data:`@TYPE@`
+variable which is preprocessed to a number of otherwise identical loops
+with several types such as :c:data:`INT, LONG, UINT, ULONG`.
```

Comment:
Those are not really type variables, they are intended as part of function names. Then actual types are strings like `npy_byte`. It is probably better to describe the operation as string substitution.